operators/src/engine/result_descriptor.rs

```diff
@@ -0,0 +1,84 @@
+use geoengine_datatypes::{
+    collections::VectorDataType, projection::ProjectionOption, raster::RasterDataType,
+};
+use serde::{Deserialize, Serialize};
+
+pub trait ResultDescriptor: Copy {
+    type DataType;
+    fn data_type(&self) -> Self::DataType;
+    fn projection(&self) -> ProjectionOption;
+}
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
+pub struct RasterResultDescriptor {
+    pub data_type: RasterDataType,
+    pub projection: ProjectionOption,
+}
+
+impl RasterResultDescriptor {
+    pub fn map<F>(self, f: F) -> Self
+    where
+        F: Fn(Self) -> Self,
+    {
+        f(self)
+    }
+
+    pub fn map_data_type<F>(mut self, f: F) -> Self
+    where
+        F: Fn(RasterDataType) -> RasterDataType,
+    {
+        self.data_type = f(self.data_type);
+        self
+    }
+}
+
+impl ResultDescriptor for RasterResultDescriptor {
+    type DataType = RasterDataType;
+    fn data_type(&self) -> Self::DataType {
+        self.data_type
+    }
+    fn projection(&self) -> ProjectionOption {
+        self.projection
+    }
+}
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
+pub struct VectorResultDescriptor {
+    pub data_type: VectorDataType,
+    pub projection: ProjectionOption,
+}
+
+impl VectorResultDescriptor {
+    pub fn map<F>(self, f: F) -> Self
+    where
+        F: Fn(Self) -> Self,
+    {
+        f(self)
+    }
+
+    pub fn map_data_type<F>(mut self, f: F) -> Self
+    where
+        F: Fn(VectorDataType) -> VectorDataType,
+    {
+        self.data_type = f(self.data_type);
+        self
+    }
+
+    pub fn map_projection<F>(mut self, f: F) -> Self
+    where
+        F: Fn(ProjectionOption) -> ProjectionOption,
+    {
+        self.projection = f(self.projection);
+        self
+    }
```

Comment:
unfortunately not that easy to put them in the trait because they modify the struct's fields

Reply:
The implementation is the same, it just lives on the trait.

Reply:
it's not possible to define a function that takes `mut self` as parameter in a trait

Reply:
You don't need to, cf. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=365a74b384263f045a7a0e5fcded26fd

Reply:
good to know, I still think it's not possible because RasterResultDescriptor's map function needs a function that maps from RasterDataType to RasterDataType and VectorResultDescriptos's map function needs a function that maps from VectorDataType to VectorDataType. Both types are enums so there is no common trait bound for defining such a function in the ResultDescriptor trait.

Reply:
8ce1d79